Extend win_psmodule

it-praktyk opened the pull request:
##### SUMMARY
The current 2.7) win_psmodule has limited functionality. Extension proposed by me adds support for almost all parameters provided by PowerShell commands used to manage modules e.g. [Install-Module](https://docs.microsoft.com/en-us/powershell/module/powershellget/install-module?view=powershell-5.1)

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_psmodule

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes -->
```paste below
ansible 2.6.4
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/<user_name>/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, Sep 21 2018, 23:26:48) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

##### ADDITIONAL INFORMATION
I'll update examples in the documentation soon.

it-praktyk committed:
The code reformated
- an indentation unified to 4 spaces
- trailing spaces removed
- aliases removed

it-praktyk committed:
The code reformated
- an indentation unified to 4 spaces
- trailing spaces removed
- aliases removed

it-praktyk committed:
Extend win_psmodule

it-praktyk committed:
Extend win_psmodule

it-praktyk commented:
CC: @dlazz

it-praktyk committed:
Correction of documentation

it-praktyk committed:
Correction of documentation

jhawkesworth commented:
@it-praktyk are you able to run the integration tests locally on your machine?
from your branch, run 
source hacking/env-setup
then run
ansible-test windows-integration win_psmodule
you will need to define some windows hosts to run against in test/integration/inventory.winrm

If the same tests run ok for you and run ok against various versions of windows server, then maybe there is a problem with the CI.

it-praktyk commented:
@jhawkesworth,
thank you for information. I'll try to run integration tests locally.

The win_psmodule requires PowerShell >= 5.0 to work correctly. Can I be sure that VMs used under CI have installed of it? Do you know what exactly VMs images are used?

Yes, but tests fail for Windows Server 2016 too :-(

jborean93 commented:
@jhawkesworth @it-praktyk that error has only been recently fixed https://github.com/ansible/ansible/pull/46378. Please rebase your PR with the latest devel changes to get the tests running again.

> The win_psmodule requires PowerShell >= 5.0 to work correctly. Can I be sure that VMs used under CI have installed of it? Do you know what exactly VMs images are used?

We test on Server 2008 (PSv3), 2008 R2 (PSv3), 2012 (PSv3), 2012 R2 (PSv4), and 2016 (PSv5). Only Server 2016 can be used on this test but the existing test should already have that logic.

it-praktyk committed:
Correction of tests

it-praktyk committed:
Correction of tests

it-praktyk commented:
@jhawkesworth, @jborean93 thank you for your help and explanations. I don't understand why but tests tasks run under CI behave differently than run "normally".

I've corrected them and now the updated win_psmodule is ready to review.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+$params = Parse-Args -arguments $args -supports_check_mode $true
 
-$name = Get-AnsibleParam -obj $params -name "name" -type "str" -failifempty $true
+$name = Get-AnsibleParam -obj $params -name "name" -type "str"
```
Why remove this as a required option? If it's about being able to remove a repo then keep this required as we would want to move repo management to another module instead.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+            "module_version" = ""
             "nuget_changed" = $false
-            "repository_changed" = $false}
+            "repository_changed" = $false
```
Let's not continue the practice here of adding a *_changed return value. We should be utilising diff for this practice instead.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
 
 $result = @{"changed" = $false
             "output" = ""
+            "module_version" = ""
```
I wouldn't return this, if we eventually decide to add support for passing in a list of packages then this wouldn't make sense anymore.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+# You can separately make operations on modules or repositories
+# but one of them is required, fails if not
+if ( [String]::IsNullOrEmpty($name) -and [String]::IsNullOrEmpty($repo) ) {
```
We should create a new module to management repositories and not lump it into this one. While it's too late to take it out we shouldn't be adding new features into `win_psmodule`

it-praktyk replied:
Resolved #48828

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
 #Requires -Module Ansible.ModuleUtils.Legacy
 
-# win_psmodule (Powershell modules Additions/Removal)
+# win_psmodule (Windows PowerShell modules Additions/Removals/Updates)
```
PowerShell is not Windows only, if we ever open this up to non-Windows host then this doesn't make sense.

it-praktyk replied:
But as I know PowerShell Core will not be supported in a visible future. Windows PowerShell is a formal name of almost all edition shipped with Windows (except 5.x in Windows server Nano (?)).

varunchopra replied:
This is not even required I think.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+    }
+}
+
+Function Install-PackageManagement {
```
So instead of having 3 separate cmdlets;

1. To update PackageManagement
2. To update PSGet
3. To install/update a package

Why don't you lump this into the 1 and share the code.

it-praktyk replied:
I've merged the code of two (Install-PackageManagement, Install-PowerShellGet).

Merging in the way how you proposed will make a maintenance much harder.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+    $PackageProvider = Get-PackageProvider -ListAvailable | Where-Object {($_.name -eq 'Nuget') -and ($_.version -ge "2.8.5.201")}
+    if (-not($PackageProvider)){
+        try{
+            Install-PackageProvider -Name NuGet -MinimumVersion 2.8.5.201 -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
```
By default EAP is set to Stop, you can remove that from the cmdlet.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+
+    if ( -not ($PackageManagement)) {
+        try {
+            Install-Module -Name PackageManagement -MinimumVersion 1.1.7.0 -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
```
EAP is already stop

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
+
+    if ( -not ($PowerShellGet) ) {
+        try {
+            Install-Module -Name PowerShellGet -MinimumVersion 1.6.0 -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
```
EAP is already stop

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,83 +1,132 @@
     }
 }
 
 Function Install-Repository {
```
Not needed for this PR but we should start to look at moving the repository actions to a separate module and start to deprecate it here.

it-praktyk replied:
Resolved by #48828

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+        'Version' = ""
+        'Versions' = @("")
+    }
+    $ExistingModule = New-Object -TypeName psobject -Property $props
```
Prefer not to deal with PSObjects, just return the hashtable if you are just setting properties

it-praktyk replied:
That proposal hurts me too much. Refused.

dagwieers replied:
We do try to avoid `PSObject` in our modules if there's no need for them.
@jborean93 I don't see it documented in https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general_windows.html#windows-new-module-development, only `PSCustomObject` is mentioned. Should we add it ?

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+        $ExistingModule.Versions = $ExistingModules.FullVersion
+
+        if ( [String]::IsNullOrEmpty($RequiredVersion) -and
```
`[String]::IsNullOrEmpty()` seems to be a bit much, you should just be able to if `if (-not ($RequiredVersion -or $MinimumVersion -or $MaximumVersion)) {`.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+            $ReturnedModule = $ExistingModules | Select-Object -First 1
+        }
+        elseif ( -not ([String]::IsNullOrEmpty($RequiredVersion)) ) {
```
`if ($RequiredVersion)`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+        elseif ( -not ([String]::IsNullOrEmpty($RequiredVersion)) ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $_.FullVersion -eq $RequiredVersion }
+        }
+        elseif ( (-not ([String]::IsNullOrEmpty($MinimumVersion))) -and
```
`elseif ($MinimumVersion -and $MaximumVersion)`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+                (-not ([String]::IsNullOrEmpty($MaximumVersion))) ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $_.Version -ge $MinimumVersion -and $_.Version -le $MaximumVersion } | Select-Object -First 1
+        }
+        elseif ( -not ([String]::IsNullOrEmpty($MinimumVersion)) ) {
```
`elseif ($MinimumVersion)`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+        elseif ( -not ([String]::IsNullOrEmpty($MinimumVersion)) ) {
+            $ReturnedModule = $ExistingModules | Where-Object -FilterScript { $_.Version -ge $MinimumVersion } | Select-Object -First 1
+        }
+        elseif ( -not ([String]::IsNullOrEmpty($MaximumVersion)) ) {
```
`elseif ($MaximumVersion)`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+                [String]::IsNullOrEmpty($MinimumVersion) -and
+                [String]::IsNullOrEmpty($MaximumVersion))  {
+
+            $ReturnedModule = $ExistingModules | Select-Object -First 1
```
Should you be selecting the first module or the latest version that's installed?

it-praktyk replied:
Returned modules are sorted from the newer version so returning the first returns the newest one.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
     }
 }
 
+Function Get-PsModule {
```
Why go through this convoluted cmdlet when you can just call [Get-InstalledModule](https://docs.microsoft.com/en-us/powershell/module/powershellget/Get-InstalledModule?view=powershell-5.1) cmdlet instead. It allows you to set the min/max/req version and returns what we need.

it-praktyk replied:
I didn't know about that cmdlet. Now I've checked how that works and 

```poweshell
PS C:\vagrant\modules> Get-Module Pester -ListAvailable


    Directory: C:\Program Files\WindowsPowerShell\Modules


ModuleType Version    Name                                ExportedCommands
---------- -------    ----                                ----------------
Script     4.4.2      Pester                              {Describe, Context, It, Should...}
Script     4.4.0      Pester                              {Describe, Context, It, Should...}
Script     3.4.0      Pester                              {Describe, Context, It, Should...}


PS C:\vagrant\modules> Get-InstalledModule Pester -RequiredVersion 3.4.0
PackageManagement\Get-Package : No match was found for the specified search criteria and module names 'Pester'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\2.0.1\PSModule.psm1:9251 char:9
+         PackageManagement\Get-Package @PSBoundParameters | Microsoft. ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...lets.GetPackage:GetPackage) [Get-Package], Exception
    + FullyQualifiedErrorId : NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.GetPackage

PS C:\vagrant\modules> [Version]$required = "3.4.0"
PS C:\vagrant\modules> Get-InstalledModule Pester -RequiredVersion $required
PackageManagement\Get-Package : No match was found for the specified search criteria and module names 'Pester'.
At C:\Program Files\WindowsPowerShell\Modules\PowerShellGet\2.0.1\PSModule.psm1:9251 char:9
+         PackageManagement\Get-Package @PSBoundParameters | Microsoft. ...
+         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ObjectNotFound: (Microsoft.Power...lets.GetPackage:GetPackage) [Get-Package], Exception
    + FullyQualifiedErrorId : NoMatchFound,Microsoft.PowerShell.PackageManagement.Cmdlets.GetPackage
```

it-praktyk replied:
So currently I leave 'this convoluted cmdlet' :-)

varunchopra replied:
You might want to check out why it doesn't work for you. This is faster and cleaner:

```
PS > Get-InstalledModule PSScriptAnalyzer -RequiredVersion 1.17.1

Version    Name                                Repository           Description
-------    ----                                ----------           -----------
1.17.1     PSScriptAnalyzer                    PSGallery            PSScriptAnalyzer provides script analysis and checks for potential code defects in t...
```

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
-        Fail-Json $result $ErrorMessage
-      }
+            $ht = @{
+                "Name" = $Name
```
There's no need to set the keys in quotes

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
-      }
+            $ht = @{
+                "Name" = $Name
+                "AllowClobber" = $AllowClobber
```
If we didn't specify `$AllowClobber` then this may not be a valid parameter.

it-praktyk replied:
Hmmm, IMHO because
```
$allow_clobber = Get-AnsibleParam -obj $params -name "allow_clobber" -type "bool" -default $false
```
the parameter will be defined always (?).

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+                "AllowClobber" = $AllowClobber
+                "SkipPublisherCheck" = $SkipPublisherCheck
+                "WhatIf" = $CheckMode
+                "ErrorAction" = "Stop"
```
EAP is already stop

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+            ForEach ($VersionParameterString in $VersionParameters) {
+                $VersionParameterVariable = Get-Variable -Name $VersionParameterString
+                if ( -not([String]::IsNullOrEmpty($VersionParameterVariable.Value)) ){
```
`if ($VersionParameterVariable.Value)`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+            $ExistingModuleAfter = Get-PsModule -Name $Name
+
+            $VersionInstalled = Compare-Object -ReferenceObject $ExistingModuleBefore.Versions -DifferenceObject $ExistingModuleAfter.Versions -PassThru
```
No need for all this once we remove the `module_version` return value.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+    if (Get-Module -Listavailable | Where-Object {$_.name -eq $Name}){
+        try{
+            $ht = @{
+                "Name" = $Name
```
No need to quote the keys

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+                "Name" = $Name
+                "Confirm" = $false
+                "WhatIf" = $CheckMode
+                "ErrorAction" = "Stop"
```
EAP is already stop

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+
+            [String[]]$VersionParameters = @("RequiredVersion","MinimumVersion","MaximumVersion")
+
+            ForEach ($VersionParameterString in $VersionParameters) {
```
You're duplicating a bit of code between here and the install part, it might be best to start sharing some of the code instead

it-praktyk replied:
The currently I left that but I hope to return to that in a near future.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -90,97 +139,346 @@ Function Remove-Repository{
+    Fail-Json $result $ErrorMessage
+}
+
+if ( -not ( [String]::IsNullOrEmpty($required_version) ) -and
```
`if ($required_version -and ($minimum_version -or $maximum_version)) {`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -11,68 +12,116 @@
 module: win_psmodule
 version_added: "2.4"
-short_description: Adds or removes a Powershell Module
+short_description: Adds or removes a Windows PowerShell Module
```
Keep as just PowerShell, you never know when/if we start opening these modules up more

it-praktyk replied:
Windows PowerShell is an official name.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -11,68 +12,116 @@
+  skip_publisher_check:
+    description:
+      - If C(yes) allows installs a newer version of a module that already exists on your computer in the case when a newer one is not
+      - digitally signed by a trusted publisher and the newest existing module is digitally signed by a trusted publisher.
```
No need for `-` on the newline if not starting a new sentence.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -11,68 +12,116 @@
+    default: no
+  skip_publisher_check:
+    description:
+      - If C(yes) allows installs a newer version of a module that already exists on your computer in the case when a newer one is not
```
`If C(yes), allows you to install a newer version of a module that already exists on your computer in the case when a newer one is not digitally signed by a trusted publisher and the newest existing module is digitally signed by a trusted publisher.`

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -82,10 +131,16 @@
-  sample: "Module PowerShellCookbook installed"
+  sample: "Module PowerShellCookbook installed."
   type: string
+module_version:
```
Let's not return this

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -96,4 +151,16 @@
   returned: always
   type: boolean
   sample: True
+powershellget_changed:
```
Let's not return this

jborean93 commented on lib/ansible/modules/windows/win_psmodule.py:
```diff
@@ -96,4 +151,16 @@
+  type: boolean
+  sample: True
+  version_added: "2.8"
+packagemanagement_changed:
```
Let's not return this

jborean93 reviewed (changes requested):
Thanks for the work, I've added a few comments to the changes. Overall I think the workflow could be streamlined a bit that's nothing major.

In the future, please try and keep formatting changes to a minimum, it makes it a lot harder to preserve who touched what line going forward. It also makes it a lot harder to review as there's more noise in diff output.

it-praktyk commented:
@jborean93, thank you for your review. My proposal is that I'll
- create a simple win_psrepository module - based at the code from the current win_psmodule
- remove the migrated code from win_psmodule in the current pull request
- include your comments in the current pull request

Ii it OK for you?

jborean93 commented:
That sounds like a good plan, I think we should try and decide on the name for the repo/source module, potential options are;

* `win_psrepository`
* `win_psrepo`
* `win_pssource`
* `win_psget_respository`
* `win_psget_repo`
* `win_psget_source`
* some other option I'm not thinking about?

If we do decide to go the `win_psget_*` option, we should also look at making an alias for this module to `win_psget_module` to they align together.

it-praktyk commented:
IMHO, the name `win_psrepository` is the best.

jborean93 commented:
Let's go for that for now, should probably discuss it at the next working group though.

pixelicous commented:
Can this be merged already? 👍 
Cant believe this wasn't a functionality by defualt, i was about to do a pull request myself :)

jborean93 commented:
@it-praktyk `win_psrepository` won out fairly quickly in the meeting so full steam ahead with that :)

@pixelicous once the review comments have been address we will definitely want to merge it in. If you can test it out and post your feedback that would be great.

it-praktyk committed:
Post-review updates and improvements

it-praktyk committed:
Post-review updates and improvements

it-praktyk committed:
Update NuGet provider at the begining of tests

it-praktyk committed:
Update NuGet provider at the begining of tests

it-praktyk committed:
Correct how/when prerequisites modules are installed

it-praktyk committed:
Correct how/when prerequisites modules are installed

it-praktyk committed:
Correct tests to be compatible with CI

it-praktyk committed:
Correct tests to be compatible with CI

it-praktyk committed:
Correct uninstalling many version, update tests

it-praktyk committed:
Correct uninstalling many version, update tests

it-praktyk committed:
Remove output, extend and correct tests

it-praktyk committed:
Remove output, extend and correct tests

it-praktyk committed:
Correct check_mode

it-praktyk committed:
Correct check_mode

it-praktyk committed:
Add integration tests for the check_mode

it-praktyk committed:
Add integration tests for the check_mode

it-praktyk committed:
Correct installing/updating NuGet

it-praktyk committed:
Correct installing/updating NuGet

Commit:
archive: Fix check if archive is created in path to be removed (#29420)

mrhockeymonkey commented:
Hi @it-praktyk great work! this is exactly what I need. Really keen to see this merged.

it-praktyk committed:
Use a hash table instead of a psobject

it-praktyk committed:
Use a hash table instead of a psobject

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
-            "output" = ""
-            "nuget_changed" = $false
-            "repository_changed" = $false}
+$result = @{changed = $false}
```
When I saying to remove the *_changed variables I was only talking about the ones that was added in this PR. Unfortunately we are stuck with the existing ones until we go through a deprecation cycle which is 4 releases. If we do want to deprecate `nuget_changed`, `repository_changed`, and `output` we should do that in a separate PR.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
+$minimum_version = Get-AnsibleParam -obj $params -name "minimum_version" -type "str"
+$maximum_version = Get-AnsibleParam -obj $params -name "maximum_version" -type "str"
 $repo = Get-AnsibleParam -obj $params -name "repository" -type "str"
-$url = Get-AnsibleParam -obj $params -name "url" -type "str"
```
We will need to go through a deprecation period (4 releases) before we can remove existing options.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
-        Fail-Json $result $ErrorMessage
-      }
+
+    # Those are minimum required versions of modules.
```
Why are these the minimum? I've been able to use this module using the versions included with PSv5 which are at `1.0.0.1`?

it-praktyk replied:
The PowerShellGet module v. 1.6.0 introduced support for prereleases.
The PackageManagement module v. 1.1.7.0 [is required by the PowerShellGet](https://github.com/PowerShell/PowerShellGet/blob/4c96ddba0c3c98318b4f546ef9f78cfb1400c514/PowerShellGet/PowerShellGet.psd1#L44) so I added that explicitly.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
+        PowerShellGet = '1.6.0'
+    }
+
+    ForEach ( $Name in $PrereqModules.Keys ) {
```
Not saying you have to but you can also use `.GetEnumerator()` like so

```
ForEach ($Name in $PrereqModules.GetEnumerator()) {
    $Name.Key
    $Name.Value
}
```

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
-      catch{
-        $ErrorMessage = "Problems removing $($Name) module: $($_.Exception.Message)"
+# Check PowerShell version, fail if < 5.0
+$PsVersion = $PSVersionTable.PSVersion
```
We should remove this check as technically older PowerShell versions can have PowerShellGet installed. Best to change this to check if PowerShellGet is installed with the required cmdlets.

jborean93 commented on lib/ansible/modules/windows/win_psmodule.ps1:
```diff
@@ -1,186 +1,380 @@
-    if (($repo) -and ($url)) {
-        Install-Repository -Name $repo -Url $url -CheckMode $check_mode 
+    if ($name) {
+        $ht = @{
```
_sigh: I wish we could do inline hashtables instead of needing to create a variable_

it-praktyk replied:
I didn't get that.

varunchopra replied:
You can just pass the parameters instead of creating a new hash table for that specific purpose.

jborean93 reviewed (changes requested):
A few things to note in additions to my comments;

* We shouldn't be removing any functionality in this PR.
    * For things we want to remove we need to add a deprecation message and the docs to indicate they will be removed in Ansible 2.x (4 versions from now)
    * To keep this PR simple, we should be doing that deprecation in another PR as it will just add noise to this one which is focused on the versioning
* We shouldn't be deleting the existing tests. Keep them there to ensure the changes in the PR do not break compatibility
    * You can run them side by side with your new tests but until these changes are settled in we shouldn't remove the older ones
* I highly recommend not using the `in_check_mode` format for the tests. I would do the check mode explicitly on the tasks themselves and follow the pattern `check test`, `test`, `idempotent test` which assertions after every one
    * It is quite difficult to follow the tests when using the `in_check_mode` format due to the numerous templating and when conditions for the assertions and knowing when they actually run
    * It can be complex when it comes to setting up scenarios for the check mode runs whereas explicitly setting `check_mode` on particular tests allows you to use the existing objects that you may have created in previous tests

Commit:
Added Ansible UTM Module for managing reverse proxy auth profile entries.

MatrixCrawler committed:
Added Ansible UTM Module for managing reverse proxy frontend entries

MatrixCrawler committed:
Added Ansible UTM Module for managing reverse proxy location entries

Commit:
- Fix sanity checks
  - Fixed documentation trailing whitespaces
  - Added missing/broken documentation

Commit:
- Fix sanity checks
  - Fixed documentation trailing whitespaces
  - Added missing/broken documentation

Commit:
- Fix sanity checks
  - Fixed documentation trailing whitespaces
  - Added missing/broken documentation

abadger committed:
Add a CLIArgs Singleton class that will hold the parse cli arguments

abadger committed:
Save the command line arguments into a global context

* Once cli args are parsed, they're constant.  So, save the parsed args
  into the global context for everyone else to use them from now on.
* Port cli scripts to use the CLIARGS in the context
* Refactor call to parse cli args into the run() method
* Fix unittests for changes to the internals of CLI arg parsing
* Port callback plugins to use context.CLIARGS
  * Got rid of the private self._options attribute
  * Use context.CLIARGS in the individual callback plugins instead.
  * Also output positional arguments in default and unixy plugins
  * Code has been simplified since we're now dealing with a dict rather
    than Optparse.Value

abadger committed:
Split up the base_parser function

The goal of breaking apart the base_parser() function is to get rid of
a bunch of conditionals and parameters in the code and, instead, make
code look like simple composition.

When splitting, a choice had to be made as to whether this would operate
by side effect (modifying a passed in parser) or side effect-free
(returning a new parser everytime).

Making a version that's side-effect-free appears to be fighting with the
optparse API (it wants to work by creating a parser object, configuring
the object, and then parsing the arguments with it) so instead, make it
clear that our helper functions are modifying the passed in parser by
(1) not returning the parser and (2) changing the function names to be
more clear that it is operating by side-effect.

Also move all of the generic optparse code, along with the argument
context classes, into a new subdirectory.

arnested committed:
Fix lastpass lookup error

Fixes #42062.

abadger committed:
Cleanups and fixes to cli

* Mark methods which are really functions as staticmethod
* Fix calls to other staticmethods to use the subclass rather than the
  base class so that any inheritance overriding will be honored.
* Remove unnecessary logic and dead code
* Fix a typo in a docstring of how to implement subclass init_parser()
  methods
* Call superclass's post_process_args in ansible-doc
* Fix copyright comment according to suggested practice

abadger committed:
Move the arguments module into cli/ and context_objects into utils

* Note: Python2 is not as intelligent at detecting false import loops as
  Python3.  context_objects.py cannot be added to cli/arguments because it
  would set up an import loop between cli/__init__.py,
  cli/arguments/context_objects.py, and context.py on Python2.

ci_complete

it-praktyk committed:
Reproduces changes from 26474c690263eb7ecda454aec07129e3303b637f

it-praktyk committed:
Reproduces changes from 26474c690263eb7ecda454aec07129e3303b637f

it-praktyk committed:
Change tests due to that Pester is digitally signed now.

it-praktyk committed:
Change tests due to that Pester is digitally signed now.

it-praktyk committed:
Revert changes to the output and the add/remove repository

it-praktyk committed:
Revert changes to the output and the add/remove repository

it-praktyk committed:
Reintroduce the older version of integration tests

it-praktyk committed:
Reintroduce the older version of integration tests

it-praktyk committed:
Change initial requirements

it-praktyk committed:
Change initial requirements

Akasurde committed:
Temporarily disable mysql_variables (#50403)

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

sdodsley committed:
Pure Storage facts fix (#50349)

* Add module for Pure Storage FlashBlade to manage directory services

* Fix facts not correctly passing into ansible_facts dict

it-praktyk committed:
Change structure of the integration tests

it-praktyk committed:
Change structure of the integration tests

it-praktyk committed:
Uncomment tests

it-praktyk committed:
Uncomment tests

abadger committed:
Cleanups to the common.sys_info API

* Move get_all_subclasses out of sys_info as it is unrelated to system
  information.
* get_all_subclasses now returns a set() instead of a list.
* Don't port get_platform to sys_info as it is deprecated.  Code using
  the common API should just use platform.system() directly.
* Rename load_platform_subclass() to get_platform_subclass and do not
  instantiate the rturned class.
* Test the compat shims in module_utils/basic.py separately from the new
  API in module_utils/common/sys_info.py and module_utils/common/_utils.py

resmo committed:
vultr_server: fix multiple ssh keys not handled (#50337)

* vultr_server: fix multiple ssh keys not handled

* add changelog fragment

felixfontein committed:
docker_secret, docker_config: allow to base64-decode data (#49688)

* Adjust docker_config and docker_secret to be more similar.

* Add data_is_b64 parameter to docker_secret and docker_config.

* Add changelog.

gundalow committed:
Disable zabbix_host 50386 (#50427)

yuwzho committed:
rename force to force_delete_nonempty (#50290)

rverchere committed:
Add support for NRDP notifications (#39297)

* Add support for NRDP notifications

* Correct pep-8 check, adding copyright

* Correct pep-8 check using autopep8

* Update nrdp to 2.6 version

* Update nrdp to 2.7 version

* Use internal ansible url and options modules

* Remove useless comments

* Add option to validate https certs

* Update nrdp to 2.8 version

* Correct nrdp pep8

abadger committed:
Fix for the new 2.8 feature to have modprobe check for builtins.

uname output includes a newline.  strip() the newline so that the
filepath we construct points to the right place.

ericzolf committed:
Fix firewalld module failing on missing protocol. (#50242)

Under Python 3.7 at least, the split of the port field fails
ungracefully if there is no slash. The fix also addresses the
case of an empty protocol after the slash.

acozine committed:
uses correct highlighting (#50457)

wojtek0806 committed:
fixed a bug in transform_name (#50213)

wojtek0806 committed:
added new module to manipulate signatre sets on ASM policy (#50209)

felixfontein committed:
Sort suboptions and subresults in docs. (#50315)

Fixes #50041.

maxamillion committed:
fix order of dnf api operations so transactions don't fail (#50038)

* fix order of dnf api operations so transactions don't fail

Previously dnf.base.fill_sack() was called before
dnf.base.update_cache() which apparently breaks dnf transaction
logic as per https://bugzilla.redhat.com/show_bug.cgi?id=1658694

Fixes #49060

Signed-off-by: Adam Miller <admiller@redhat.com>

* add changelog and test case

Signed-off-by: Adam Miller <admiller@redhat.com>

maxamillion committed:
add changelog for #50242 (#50480)

Signed-off-by: Adam Miller <admiller@redhat.com>

jbenden committed:
Add support for Kali Linux detection. (#50331)

Signed-off-by: Joseph Benden <joe@benden.us>

smueller18 committed:
disable docker_network ipam tests (#50477)

felixfontein committed:
docker_swarm_service: Fix typo for option type (#50483)

* Fix typo for option type in docker_swarm_service.

* Add changelog.

felixfontein committed:
openssl_csr: idempotency doesn't work correctly for keyUsage (#50361)

* Fix key usage idempotency bug.

* Extend tests.

* Add changelog.

felixfontein committed:
openssl_*: prevent error when path includes no path (#50322)

* Prevent error when path includes no path.

* Add changelog.

lolcube committed:
add openssh_cert module (#49605)

* add openssh_cert module

* fix some linter issues

* address review feedback

* add basis for integration tests

* change valid datetime formats

* handle principals as set not as list

* handle principals as set not as list

* handle principals as set not as list

it-praktyk committed:
Add choices of the state parameter documentation (#50481)

* Add choices of the state parameter documentation

* Consistency

machacekondra committed:
ovirt_host: Improve failed upgrade check (#50489)

fretis committed:
Update playbooks_loops.rst (#50437)

Error in the example for exclusion pattern

+label: docsite_pr

Akasurde committed:
Update changed_when and failed_when examples (#50411)

Added examples in playbooks_error_handling doc for handlining
multiple conditions in changed_when and failed_when

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde committed:
gcp: documentation update (#50183)

* gcp: documentation update

* Update example about dynamic inventory
* minor typo fixes in gcp_utils
* Additional information about enabling inventory plugin in ansible.cfg

partially fixes: #44404

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

richlv committed:
Update zabbix_action.py (#50490)

* Add trailing dots for consistency.
* Document how the current host can be specified for "run_on_hosts".

+label: docsite_pr

Commit:
Fixes #37209 - Added complex return value, instance (#49989)

felixfontein committed:
Docs: add Pygments lexer for Ansible output (#50318)

* Add specialized lexer for Ansible output.

* Make linter happy.

* Use different tokens.

Commit:
Make WinRM security warning more explicit (#50263)

Make WinRM security warning more explicit.
Currently the warning is "viewed by anyone", this is a much lower risk than arbitrary command injection. Therefore the risk should be phrased appropriately.

+label: docsite_pr

dagwieers committed:
Windows: Add missing parameter types and doc fixes (#50232)

* Windows: Add missing parameter types and doc fixes

This PR includes:
- Adding missing parameter types
- Various documentation fixes

* Update lib/ansible/modules/windows/win_copy.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_credential.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_domain_computer.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_domain_user.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_environment.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_psexec.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_uri.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Update lib/ansible/modules/windows/win_wait_for.py

Co-Authored-By: dagwieers <dag@wieers.com>

* Ensure docstrings are raw strings

drewmullen committed:
allow for vault enterprise namespaces (#50462)

* enable namespaces feature for hashi_vault lookup

* include version_added dict in options documentation

jdorel committed:
Add galaxy.html#dependencies link (#49984)

* Add galaxy.html#dependencies link

+label: docsite_pr

s-hertel committed:
Add unstable alias to s3_bucket integration tests (#50508)

jborean93 committed:
dnf - fix conf_file loading (#50515)

jborean93 committed:
ansible test: re-enable mysql tests with fix (#50430)

sivel committed:
Give IncludedFile more context via ansible_search_path (#50045)

* Give IncludedFile more context via ansible_search_path to template lookups. Fixes #49969

* Update units

dagwieers committed:
urldecode filter for Jinja2 (#28503)

* urldecode filter for Jinja2

We needed this in order to deconstruct correct URLs using Jinja2.
And we might as well upstream this.

* Add integration tests

* Fixes for Python 3

* Add urlencode for older Jinja2

dagwieers committed:
ACI: Implement encoded query_string (#50358)

This requires urldecode support in Ansible

jborean93 committed:
ansible-test: docker_network set as untable (#50526)

* Revert "disable docker_network ipam tests (#50477)"

This reverts commit 4dd2b8e2d004418a981ff7c9552792d44f19131a.

* test - docker_network mark as unstable

richlv committed:
Update documentation for mysql_db notes (#50444)

* Simplify the requirement section to avoid duplicating info already in mysql_document_fragment.
* Package installation section simplified, mentioned RHEL; a typo fix in "than".
* added dnf for Fedora

Akasurde committed:
Handle exception raised in recursive_finder API (#49590)

User module can contain Indentation errors or syntax errors.
Handle AST exceptions rather than showing traceback while importing such module.

Fixes: #21707

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

justjais committed:
Added new module to support NIOS Fixedaddress DHCP property (#49119)

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

* nios fixedaddr new module

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

* updating name

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

* added unit tests

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

Commit:
remove forced lowercase on group_name (#50445)

yungezz committed:
azure_rm_webapp: fix updating app settings (#49181)

unrinfosec committed:
Correct Errors in Documentation (#50507)

* Correct Errors in Documentation

According to the module, "compose" is not required, but the docs said it was - removed the "required" tag. Additionally, the example for removing a stack did not include the required "name" option (this method is why "compose" is not required.
+label: docsite_pr

* Removed change to description

Corrected issue to my "correction" on the description.

* Removed required false

* Removed all instances of required: false.

ganeshrn committed:
Fix backup issue in network config modules and network action plugins common code refactor (#50301)

* Fix backup issue in network config modules

*  Fix `get_working_path` not found issue introduced due to
   backup config code refactor (PR #50208)

*  Further refactor config related action plugins to minimize
   duplicate code

*  Remove unwated imports in config action plugins

* Add common network class for action plugin and related code refactor

* Fix review comment

trishnaguha committed:
add privileged role validation for nxos become (#50312)

Signed-off-by: Trishna Guha <trishnaguha17@gmail.com>

ganeshrn committed:
Fix junos terminal error regex (#50538)

*  Change junos terminal error regex to read error string
   correctly.

jborean93 committed:
dnf/yum - added install_weak_deps option (#50525)

* dnf/yum - added install_weak_deps option

* skip creating weak dep packages if rpm doesn't support it

* fix rpm check for older hosts

abadger committed:
Add a porting guide entry for ansible_distribution facts (#50251)

* Add a porting guide entry for ansible_distribution facts

Switching away from platform.distro() will cause changes sometimes due
to the new code using new sources of information that may be out of sync
with the old ones.  Just have to make people aware of that and also what
we are doing to mitigate it when appropriate.

* wordsmithed, added links for new distro backend

wojtek0806 committed:
adds apm policy fetch module (#50534)

wojtek0806 committed:
Changes exists() method logic in asm policy fetch, corrects documentation (#50560)

wojtek0806 committed:
adds bigip apm policy import (#50559)

wojtek0806 committed:
Corrects doc strings in bigip_asm_policy_* modules (#50561)

sivel committed:
Add new AnsibleTemplateError to more easily catch templating issues (#50563)

* Add new AnsibleTemplateError to more easily catch templating issues. Fixes #50154

* Add changelog fragment

sivel committed:
module_utils/urls.py is core (#50567)

* module_utils/urls.py is core

* More core supported module_utils

* More missed core module_utils

kairoaraujo committed:
new module: Configure LVM and NFS file systems for AIX (#30810)

* new module: AIX LVM file system and NFS

This module creates, removes, mount and unmount LVM and NFS file system
for AIX using /etc/filesystems. For LVM file systems is also possible
to resize the file system.

* better parameters options structure

better parameters options structure

* Improved file system resize returns

Added better tratment for return codes for file system resize.
When a resize is not possible because no enough space on lv or
shrink is not allowed.

* improved doc and creation file system return code

- improved doc
- creation file system return code 10 was treated.

* Doc recomendations, dict result, line limit

- Added doc recomendations
- Changed return to dict results on main()
- Using 159 columns for code limit

* wrong changed return when file system is already

Fixed wrong changed return when file system is already mounted.
When the file system is already mounted the return for changed
is False.

* Fixed description and included playbook for tests

- Fixed description
- Included playbook for manual tests

* Various small bits to get this merged ASAP

* Rename test/legacy/aix_filesystem.yml to test/integration/targets/aix_filesystem/tasks/main.yml

Move integration test to its proper location

* Create aliases

* Fix CI issues

varunchopra reviewed (commented):
There's a lot going on here but I think `Install-PackageProvider` should be deprecated from this seeing how we already have a separated module for installing it.

Just because someone has specified a repo along with the module with `state = "absent"` shouldn't warrant it's removal and will behave unexpectedly. 

Overall I think it's a great improvement over the old module which would've given me a stroke.

richlv committed:
Improve style and fix typos in the Zabbix module titles and descriptions (#50575)

spoyd committed:
(action/script) Fix windows absolute path detection when executing powershell script (#50365)

rajaspachipulusu17 committed:
Pluribus Networks module for connection stat settings with UT's (#50186)

* Pluribus Networks module for connection stat settings

Commit:
adding module azure_rm_virtualmachinescalesetinstance_facts (#50440)

mkrizek committed:
Fix searchpath in the template lookup (#50488)

rcarrillocruz committed:
Checkpoint httpapi plugin (#49929)

* Add checkpoint httpapi plugin and access rule facts module

* WIP checkpoint_access_rule module

* Add publish and install policy, plus fix empty json object request for publish

* Refactor publish and install_policy onto module_utils

* Add update resource logic

* Add checkpoint_host_facts module

* Return code and response on get_acess_rule function

* Add checkpoint_host module

* Add checkpoint_run_script module

* Add checkpoint_task_facts module

* Show all tasks if no task id is passed

Note, this is only available on v1.3 of Checkpoint WS API

* Add update logic to checkpoint host

* Add full details on get task call

* Add checkpoint httpapi plugin

* Fix pep8

* Use auth instead of sid property and return False on handle_httperror method

* Fix version in docstring

* Remove constructor

* Remove Accept from base headers

* Do not override http error handler and assign Checkpoint sid to connection _auth

There is scaffolding in the base class to autoappend the token, given
it is assigned to connection _send

* Use new connection queue message method instead of display

* Remove unused display

* Catch ValueError, since it's a parent of JSONDecodeError

* Make static methods that are not used outside the class regular methods

* Add missing self to previously static methods

* Fix logout

Was carrying copy pasta from ftd plugin

* Remove send_auth_request

* Use BASE_HEADERS constant

* Simplify copyright header on httpapi plugin

* Remove access rule module

* Remove unused imports

* Add unit test

* Fix pep8

* Add test

* Add test

* Fix pep8

mkrizek committed:
Allow for vaulted templates in template lookup (#49819)

Fixes #34209

dagwieers committed:
raw is transport-agnostic

+label: docsite_pr

abadger committed:
Clarify the change made to reverse_inventory

Qalthos committed:
Copy detailed logging to persistent.py so local can benefit, too. (#50614)

dagwieers committed:
xml: Improve documentation and examples (#50602)

This PR includes:
- Type information for parameters
- Added 'seealso' section to documentation
- Improve examples

devjatkin committed:
add env variable VAULT_AUTH_METHOD (#50572)

it-praktyk committed:
Make the module and integration tests compatible with PowerShell >= 3.x

it-praktyk committed:
Make the module and integration tests compatible with PowerShell >= 3.x

it-praktyk committed:
Reproduces changes from d8630271598c4014e224348416cf229f2f31fdf7

it-praktyk committed:
Resolve post-rebasing conflicts

it-praktyk commented:
I close that pull request now because I did a mess when I tried to rebase.
